[pytorch] Lazy Module Fix for kwargs argument support on forward

ertrue opened the pull request:
Summary:
This is a fix to resolve the issue: #35643 and #59923

Test Plan:
python3 test: caffe2/test/test_nn.py

Differential Revision: D29687148

albanD commented on torch/nn/modules/lazy.py:
```diff
@@ -174,7 +174,7 @@ def __init__(self: _LazyProtocol, *args, **kwargs):
         super().__init__(*args, **kwargs)  # type: ignore[misc]
         self._load_hook = self._register_load_state_dict_pre_hook(self._lazy_load_hook)
-        self._initialize_hook = self.register_forward_pre_hook(self._infer_parameters)
+        self._initialize_hook = self.register_forward_pre_hook(self._infer_parameters, with_kwargs=True)
```
This would be BC-breaking no?
Even for simple things like the LazyLinear, this will make it fail if any argument is passed as a kwargs.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -26,6 +26,24 @@ def __repr__(self):
+        if self.with_kwargs:
+            return self.hook(module, input, kwargs)
+        else:
+            return self.hook(module, input)
```
You will need to fix the return type here to make sure both the updated input and kwargs are returned.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1059,8 +1077,8 @@ def _call_impl(self, *input, **kwargs):
-            for hook in (*_global_forward_pre_hooks.values(), *self._forward_pre_hooks.values()):
-                result = hook(self, input)
+            for hook_wrapper in (*_global_forward_pre_hooks.values(), *self._forward_pre_hooks.values()):
+                result = hook_wrapper.run(self, input, kwargs)
```
This should allow to update the kwargs now

Commit:
[pytorch] Lazy Module Fix for kwargs argument support on forward (#61606)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61606

This is a fix to resolve the issue: https://github.com/pytorch/pytorch/issues/35643 and https://github.com/pytorch/pytorch/issues/59923

Test Plan:
python3 test: caffe2/test/test_nn.py

python3 test on torchrec/fb/module

Differential Revision: D29687148

fbshipit-source-id: 5602a79e7a831945bc9344574593b9e0ed212ddf

y-x-c commented:
hi team, is there a plan to support this feature in the future? Thanks.
